zonemaster/zonemaster-gui
# Domain test form update

## Purpose

Improve form accessibility

## Context

2022-11-08 team meeting 

## Changes

* move protocol and profile in the same section, at the end of the form
* use `fieldset` to delimit sections of form
* add aria-label for input that do not have a visible label
* add titles for button
* remove gradient and adjust color contrast
* remove background for "about" section
* only display profile select if several profiles are available
* use aria-describedby to link validation error to inputs
* keep at least one empty row when fetching ds from parent zone and no ds are found
* fetch data for DS and NS independently
* fix "Run test" button disabled when no data is found in parent zone
* disable the whole form instead of just the domain input and check button when waiting for a rpc response

![Screenshot 2022-11-28 at 17-02-08 Zonemaster](https://user-images.githubusercontent.com/19394895/204324319-b4ed7c42-4aa1-4078-8955-982ccfabeed5.png)

## How to test this PR

Check that the form UI visually matches the screenshot in Changes.
Check that running a test is working, the progress bar is displayed and working.

Test fix of empty response for fetch data
* For a non existing domain (e.g. `toto`) and a domain without DS (e.g. `toto.fr`)
  * Try to fetch data (NS and DS) for the domain
  * Check that a warning notification is displayed for DS (and NS for a non existing domain) but the buttons / input are not disabled
  * Check the DS (and NS for a non existing domain) forms have one empty row


Test form disabled
* Try to fetch data (NS and DS) for a existing domain
* Check that the form is disabled during the time of the query
* Check that a info notification is displayed

Test profile select only displayed if more than one choice
* Make sure to have only one profile in backend
* Load the "Run test" page
* Check that the "profile" select is not displayed
* Add a second profile in backend and restart
* Reload the "Run test" page
* Check that the "profile" select is displayed with a label, aligned with the "ipv4/ipv6" checkbox and contains the two profiles

Test aria descriptions
* Do once using empty inputs to test "required" errors and once by filling form with garbage to test "server" errors
  * Generate form errors.
  * Using [Firefox accessibility inspector](https://firefox-source-docs.mozilla.org/devtools-user/accessibility_inspector/index.html) check that each input with an error has an associated description.
    ![image](https://user-images.githubusercontent.com/19394895/204327893-7b3ad7e6-ba2f-4442-b903-89ad2b63b7fb.png)

Test aria labels
* Using the Firefox accessibility inspector again check that all buttons and inputs have a textual name.
   ![image](https://user-images.githubusercontent.com/19394895/204328806-e27c5297-1824-40ec-a00e-f078cce6a6a9.png)

Test error on fetch data
* Stop the backend
* Try to query for parent data for any domain
* Check that an error notification is displayed
* Check that the form is not disabled
